### ISSUE TITLE
Changes on a alias datapoint require the restart of instances to take effect

### ISSUE THREAD

**Issue by Xyolyp** [ioBroker/ioBroker.js-controller#659]
After changing the definition of a alias datapoint I mist restart all instaces that read or weite the datapoint for the changes to take effect. In my imagination the alias datapoints should bot behave differently than any other datapoint from the point of view of an instance or a Adapter developer respectively.

**Comment by Apollon77** [comment]
Do you have and reproducable example?

**Comment by Xyolyp** [comment]
Any datapoint with enabled sql history requires me to restart the sql instance. 
If you use the datapoint below and change the read function the sql Adapter uses the old read function until it is restarted. The sql Adapter is within a compact group just in case that makes a difference. Another example of this is that I must restart the script instance after any change and after adding a new aliasdatapoint.
If you need any further data I'll happily provide what you need to resolve this issue.
```
{
  "from": "system.adapter.admin.0",
  "user": "system.user.admin",
  "ts": 1581257478419,
  "common": {
    "name": "Bett.Jonathan",
    "role": "",
    "type": "boolean",
    "desc": "Manuell erzeugt",
    "read": true,
    "write": false,
    "def": false,
    "alias": {
      "id": "sonoff.0.BedJonathan.ANALOG_A0",
      "read": "val < 743"
    },
    "custom": {
      "sql.0": {
        "enabled": true,
        "changesOnly": true,
        "debounce": 0,
        "maxLength": 10,
        "retention": "31536000",
        "changesRelogInterval": 0,
        "changesMinDelta": 0,
        "storageType": "",
        "aliasId": ""
      }
    }
  },
  "native": {},
  "acl": {
    "object": 1636,
    "owner": "system.user.admin",
    "ownerGroup": "system.group.administrator",
    "state": 1636
  },
  "_id": "alias.0.Schlafzimmer.Bett.Jonathan",
  "type": "state"
}
```

**Comment by foxriver76** [comment]
I am not able to reproduce this. But I can imagine, that the user thinks that this is the problem.
When adding a new alias via Admin, Admin does not show the value instantly, instead, it will only update the value if you reload the objects tab, e. g. by navigating to instances (what you might have done to restart the adapter) and when navigating back to objects tab, the state is shown correctly. (so no restart required) However, after creating a new alias this appeared, but when checking via cli, the state had the correct value.

In summary, it seems, that the bug is in the admin adapter.

**Comment by Apollon77** [comment]
But maybe 8especially from the last thing with changing read/write function also adapter.js L1519ff could be reason because read and write functions do not seems to be adjusted

**Comment by Apollon77** [comment]
hm ... wait ... no alias substruct is copied over ...

**Comment by Apollon77** [comment]
Hm ... on object change in adapter.js L1506 the alias object is read and the targets are checked and pot. adjusted. but the "aliasses[id].source" is not changed/updated.

For state changes this.aliases[id].source is used to use the write function ... so yes there seems to be some cases missing when objects gets updated ...
@GermanBluefox 
Maybe really always do a "remove/addAlias Subscriptions to get everything fresh initialized?

**Comment by Apollon77** [comment]
ok, write function is never used from sources ... but still question is why sources do not get updated ?! Maybe not needed?

**Comment by foxriver76** [comment]
> I am not able to reproduce this. But I can imagine, that the user thinks that this is the problem.
> When adding a new alias via Admin, Admin does not show the value instantly, instead, it will only update the value if you reload the objects tab, e. g. by navigating to instances (what you might have done to restart the adapter) and when navigating back to objects tab, the state is shown correctly. (so no restart required) However, after creating a new alias this appeared, but when checking via cli, the state had the correct value.
> 
> In summary, it seems, that the bug is in the admin adapter. @Xyolyp can you confirm this behaviour?

@Xyolyp would still be interesting if this is the same behaviour which you have experienced?

**Comment by Xyolyp** [comment]
@foxriver76 a reload of the page is sufficient to get the correct values to display in the admin adapter. But much worse: all other adapters (e.g. script, habpanel, ...) still require a restart for the changes to take effect. I don't mind reloading a page for a few months but restarting the ioBroker to be sure that the changes take effect is more annoying especially to anyone else in the house who is expecting the lights to work and all that.

**Comment by Apollon77** [comment]
Can you povide us a "as simple as possible step by step info" on how to reproduce your behaviour? And please verify that you use js-controller 2.2.9

**Comment by foxriver76** [comment]
Minimum working example - assume you write a dummy adapter or enhance logic of existing, to subscribe to all foreign states and log the state change:

- now create an alias via admin
- set the state of the alias dp
- the state change of the adapter is not triggered
- if you do the same but instead of an alias you add a new object to e. g. javascript.0 namespace via admin, the state change event is fired..

**Comment by Apollon77** [comment]
I think I have an idea ... How all of this alias stuff in adapter.js works ... basically nothing works automatically, BUT is checked and initialized on subscribe (!!) and used in get/setSate (which is irrelevant here).

Means: when you subscribe to an alias then this is handled and all aliasses that exaists are collected and subscribed and such. When you now add a new alias object that would match to the subscribe-pattern all of this "subscribe handling" is NOT done!

So to fix this we would need to remember all alias level subscriptions (starts with alias.0) and then on a new object check if one of these patterns match here and then "run the same code as it would have been done on initial subscribe" for this object ...

**Comment by Xyolyp** [comment]
I can reproduce the issue in the same way but instead of with new aliases I
can reproduce it by editing an existing alias.

Ingo Fischer <notifications@github.com> schrieb am Di. 17. März 2020 um
18:28:

**Comment by Apollon77** [comment]
And what exactly you edit= read/write function as described above or something different?

**Comment by Xyolyp** [comment]
Just The read or write function or the target data point.

Ingo Fischer <notifications@github.com> schrieb am Mi. 18. März 2020 um
22:32:

**Comment by foxriver76** [comment]
I will test this, but it should work now.

**Comment by foxriver76** [comment]
@Xyolyp Tried your scenarios with my PR branch (https://github.com/ioBroker/ioBroker.js-controller/pull/729). Everything works, thus it will be fixed with js-c 3

**Comment by Xyolyp** [comment]
I'll have a look at it on the weekend when I have time to make a rollback if necessary.

**Comment by Xyolyp** [comment]
@foxriver76 what is the easiest way to install the current state of 3.0 to test if the changes have the desired effect?

**Comment by Apollon77** [comment]
@Xyolyp The master from js-controller can always be installed via "npm install ioBroker/ioBroker.js-controller"

**Comment by Xyolyp** [comment]
@Apollon77 the changes you've made seem to fix the issues I was having. I'll go back to a more stable version of the js-controller and wait for the stable repo to receive 3.x. Thank you
